### ISSUE TITLE
Latest molecule image fails at destroy

### ISSUE THREAD

**Issue by ericlake** [ansible/molecule#1660]
# Issue Type

- Bug report

# Molecule and Ansible details

This issue is in the latest docker image of molecule.
```
quay.io/ansible/molecule                                    latest              502e06e6a9ac        4 hours ago         499 MB
```

Molecule installation method (one of):

- docker

Ansible installation method (one of):

- OS package

Detail any linters or test runners used:

Our molecule.yml file

```yaml
---
dependency:
  name: galaxy

driver:
  name: docker

lint:
  name: yamllint

platforms:
  - name: instance
    image: "geerlingguy/docker-${MOLECULE_DISTRO:-centos7}-ansible:latest"
    command: ${MOLECULE_DOCKER_COMMAND:-""}
    volumes:
      - /sys/fs/cgroup:/sys/fs/cgroup:ro
    privileged: true
    pre_build_image: true

provisioner:
  name: ansible
  lint:
    name: ansible-lint
  playbooks:
    converge: ${MOLECULE_PLAYBOOK:-playbook.yml}

scenario:
  name: default
  test_sequence:
    - lint
    - destroy
    # - dependency
    - syntax
    - create
    # - prepare
    - converge
    - idempotence
    # - side_effect
    - verify
    - destroy

verifier:
  name: testinfra
  options:
    vv: true
  lint:
    name: flake8
```

# Desired Behavior

The desired behavior is that a `molecule test` would succeed on the initial destroy operation.

# Actual Behaviour

Output from running `molecule test`

```
$ docker run --rm -v $(pwd):/tmp/$(basename "${PWD}"):z -v /var/run/docker.sock:/var/run/docker.sock -w /tmp/$(basename "${PWD}") quay.io/ansible/molecule:latest sudo molecule test
--> Validating schema /tmp/docker/molecule/default/molecule.yml.
Validation completed successfully.
--> Test matrix
    
└── default
    ├── lint
    ├── destroy
    ├── syntax
    ├── create
    ├── converge
    ├── idempotence
    ├── verify
    └── destroy
    
--> Scenario: 'default'
--> Action: 'lint'
--> Executing Yamllint on files found in /tmp/docker/...
Lint completed successfully.
--> Executing Flake8 on files found in /tmp/docker/molecule/default/tests/...
Lint completed successfully.
--> Executing Ansible Lint on /tmp/docker/molecule/default/playbook.yml...
Lint completed successfully.
--> Scenario: 'default'
--> Action: 'destroy'
    
    PLAY [Destroy] *****************************************************************
    
    TASK [Destroy molecule instance(s)] ********************************************
    changed: [localhost] => (item=None)
    changed: [localhost]
    
    TASK [Wait for instance(s) deletion to complete] *******************************
    failed: [localhost] (item=None) => {"attempts": 1, "censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result", "changed": false}
    fatal: [localhost]: FAILED! => {"censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result", "changed": false}
    
    PLAY RECAP *********************************************************************
    localhost                  : ok=1    changed=1    unreachable=0    failed=1
    
    
ERROR: 
An error occurred during the test sequence action: 'destroy'. Cleaning up.
--> Scenario: 'default'
--> Action: 'destroy'
    
    PLAY [Destroy] *****************************************************************
    
    TASK [Destroy molecule instance(s)] ********************************************
    changed: [localhost] => (item=None)
    changed: [localhost]
    
    TASK [Wait for instance(s) deletion to complete] *******************************
    failed: [localhost] (item=None) => {"attempts": 1, "censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result", "changed": false}
    fatal: [localhost]: FAILED! => {"censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result", "changed": false}
    
    PLAY RECAP *********************************************************************
    localhost                  : ok=1    changed=1    unreachable=0    failed=1
    
    
ERROR: 
```

This does not occur with an image from 3 weeks ago with the hash of `e72db41d7ece`. See the below snippet.

```
$ docker run --rm -v $(pwd):/tmp/$(basename "${PWD}"):z -v /var/run/docker.sock:/var/run/docker.sock -w /tmp/$(basename "${PWD}") quay.io/ansible/molecule:e72db41d7ece sudo molecule test
--> Validating schema /tmp/docker/molecule/default/molecule.yml.
Validation completed successfully.
--> Test matrix
    
└── default
    ├── lint
    ├── destroy
    ├── syntax
    ├── create
    ├── converge
    ├── idempotence
    ├── verify
    └── destroy
    
--> Scenario: 'default'
--> Action: 'lint'
--> Executing Yamllint on files found in /tmp/docker/...
Lint completed successfully.
--> Executing Flake8 on files found in /tmp/docker/molecule/default/tests/...
Lint completed successfully.
--> Executing Ansible Lint on /tmp/docker/molecule/default/playbook.yml...
Lint completed successfully.
--> Scenario: 'default'
--> Action: 'destroy'
    
    PLAY [Destroy] *****************************************************************
    
    TASK [Destroy molecule instance(s)] ********************************************
    changed: [localhost] => (item=None)
    changed: [localhost]
    
    TASK [Wait for instance(s) deletion to complete] *******************************
    ok: [localhost] => (item=None)
    ok: [localhost]
    
    TASK [Delete docker network(s)] ************************************************
    
    PLAY RECAP *********************************************************************
    localhost                  : ok=2    changed=1    unreachable=0    failed=0
    
    
--> Scenario: 'default'
--> Action: 'syntax'
```

**Comment by abeluck** [comment]
I too am seeing this issue, but I have logging enabled, and the error is:

```"msg": "Failed to import docker or docker-py - No module named docker. Try `pip install docker` or `pip install docker-py` (Python 2.6)"
```
Adding `docker` to the pip packages variable in the Dockerfile fixes this.

Even though `molecule[docker]` is one of the extras, the actual docker python library is not installed.

**Comment by ericlake** [comment]
When running a quick test of the pip install command used in the Dockerfile in a virtualenv I see the following.

```bash
$ PIP_PACKAGES=molecule[azure,docker,docs,ec2,gce,lxc,openstack,vagrant,windows]

$ pip install --no-cache-dir ${PIP_PACKAGES}
Collecting molecule[azure,docker,docs,ec2,gce,lxc,openstack,vagrant,windows]
  Downloading https://files.pythonhosted.org/packages/84/97/e5764079cb7942d0fa68b832cb9948274abb42b72d9b7fe4a214e7943786/molecule-2.19.0-py3-none-any.whl (180kB)
    100% |████████████████████████████████| 184kB 9.2MB/s 
  molecule 2.19.0 does not provide the extra 'azure'
  molecule 2.19.0 does not provide the extra 'docker'
  molecule 2.19.0 does not provide the extra 'docs'
  molecule 2.19.0 does not provide the extra 'ec2'
  molecule 2.19.0 does not provide the extra 'gce'
  molecule 2.19.0 does not provide the extra 'lxc'
  molecule 2.19.0 does not provide the extra 'openstack'
  molecule 2.19.0 does not provide the extra 'vagrant'
  molecule 2.19.0 does not provide the extra 'windows'
Collecting python-gilt==1.2.1 (from molecule[azure,docker,docs,ec2,gce,lxc,openstack,vagrant,windows])
  Downloading https://files.pythonhosted.org/packages/d5/8b/aafb0d991763df480dc149b8c034b57ea42d7de4ef8d5b5137c6f38105a6/python_gilt-1.2.1-py3-none-any.whl
Collecting Jinja2==2.10 (from molecule[azure,docker,docs,ec2,gce,lxc,openstack,vagrant,windows])
  Downloading https://files.pythonhosted.org/packages/7f/ff/ae64bacdfc95f27a016a7bed8e8686763ba4d277a78ca76f32659220a731/Jinja2-2.10-py2.py3-none-any.whl (126kB)
    100% |████████████████████████████████| 133kB 20.8MB/s 
```

Notice the `molecule 2.19.0 does not provide the extra 'docker'`. Once pip is done installing a `pip freeze` does not list docker.

**Comment by ragingpastry** [comment]
It looks like this was changed here: https://github.com/ansible/molecule/commit/f598cb44bc23869c533da1e1294930b91196230b#diff-380c6a8ebbbce17d55d50ef17d3cf906. I'm guessing documentation was updated but a new package wasn't pushed to pypi with the changes.

This should probably stay open until a new release tagged and put on pypi, but this should be resolved in master.

**Comment by tace** [comment]
Started first time to explore molecule and immediately experienced this. Just run simple first test by using docker image "quay.io/ansible/molecule:latest" according to this https://molecule.readthedocs.io/en/latest/examples.html.

**Comment by lwm** [comment]
Sorry to hear you had a rough start @tace! We're working on getting a shorter release schedule (a part of why this issue occurred) and keeping the Docker containers fresh and building ASAP.

**Comment by tace** [comment]
Not so much for shorted release schedule but mostly would like to see some stable versions (one or two per year) in here:
https://quay.io/repository/ansible/molecule?tab=tags 
And would like to see/select exact version of ansible and molecule on that image. As we need to fix at least ansible version exactly to match ansible version used in production.

**Comment by ericlake** [comment]
I noticed that a new version of latest was cut however the issue still remains.

**Comment by tace** [comment]
I have been fine by using 2.19 image so far. Until now I hit this one https://github.com/ansible/molecule/pull/1615 which will be fixed in next release.

**Comment by tace** [comment]
sorry I hit this https://github.com/ansible/molecule/issues/1614 problem and need #1615 to overcome it.

I use docker images with ssh and systemd installed (https://github.com/chrismeyersfsu/provision_docker/tree/master/files).

It would be nice to have:
 1. support for real ssh in molecule docker driver (instead of "docker exec")
 2. molecule provided Dockerfiles/registry for node images having ssh/systemd installed.

**Comment by apollo13** [comment]
> 2. molecule provided Dockerfiles/registry for node images having ssh/systemd installed.

Why? The workflow for molecule usually is that you let molecule create docker containers to test against. Those docker containers (depending on your tests etc) need systemd, but the container running molecule itself does usually not.

**Comment by tace** [comment]
for the point 2 I ment docker images to be tested, not docker image of molecule (this ticket).

**Comment by apollo13** [comment]
Ah okay, but why would molecule provide them? Seems a little bit out of scope. That said I agree that a set of curated and updated base images might go along way -- just not sure if that would be part of molecule's responsibilities.

**Comment by alikulov** [comment]
> Adding `docker` to the pip packages variable in the Dockerfile fixes this.
@abeluck could you please provide more details how did you solve the issue.

**Comment by abeluck** [comment]
In the Dockerfile, find the PIP_PACKAGES line

```
ENV PIP_PACKAGES="\
    molecule[azure,docker,docs,ec2,gce,lxc,openstack,vagrant,windows] \
    virtualenv \
"
```

and add `docker` to it:

```
ENV PIP_PACKAGES="\
    molecule[azure,docker,docs,ec2,gce,lxc,openstack,vagrant,windows] \
    virtualenv docker \
"
```

**Comment by decentral1se** [comment]
OK, we've made a new image release on https://quay.io/repository/ansible/molecule?tab=tags. Please check it out and let us know if there are issues. We also had some solid work done to upgrade the build to match the new package configuration: https://github.com/ansible/molecule/pull/1797. Closing this one off :+1:
